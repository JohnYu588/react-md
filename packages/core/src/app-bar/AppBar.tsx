import { cnb } from "cnbuilder";
import type { ElementType, HTMLAttributes, Ref } from "react";
import { forwardRef } from "react";
import type { CssPosition } from "../types.js";
import { bem } from "../utils/bem.js";

declare module "react" {
  interface CSSProperties {
    "--rmd-app-bar-height"?: string | number;
    "--rmd-app-bar-background-color"?: string;
    "--rmd-app-bar-color"?: string;
    "--rmd-app-bar-surface-background-color"?: string;
    "--rmd-app-bar-surface-color"?: string;
  }
}

const styles = bem("rmd-app-bar");

export type AppBarPosition = "top" | "bottom";

/**
 * - `"clear"` - the background color will be transparent
 * - `"primary"` - defaults to the current primary theme color
 * - `"secondary"` - defaults to the current secondary theme color
 * - `"surface"` - this will use the current surface background color which
 *   defaults to `#fff` for light themes, `#242424` for dark themes, and
 *   `#424242` for dark themes when the `$disable-dark-elevation` is set to
 *   `true`
 *
 * The default dark theme surface color also depends on the `$fixed-elevation`
 * value.
 */
export type AppBarTheme = "clear" | "primary" | "secondary" | "surface";

/**
 * - `"auto"` - the height will be determined by the height of the content
 * - `"normal"` - defaults to `3.5rem` (`$height`)
 * - `"prominent"` - defaults to `7rem` (`$prominent-height`)
 * - `"dense"` - defaults to `3rem` (`$dense-height`)
 * - `"prominent-dense"` - defaults to `6rem` (`$prominent-dense-height`)
 */
export type AppBarHeight =
  | "auto"
  | "normal"
  | "prominent"
  | "dense"
  | "prominent-dense";

/** @remarks \@since 6.0.0 */
export interface AppBarClassNameOptions {
  className?: string;

  /**
   * Set this to `"fixed"` or `"sticky"` to set `position: fixed;` or
   * `position: sticky;` to the app bar. The default position will be static and
   * inline with other content.
   *
   * @defaultValue `'static'`
   */
  position?: CssPosition;

  /**
   * The position within the page to "fix" the `AppBar` when the `fixed` prop is
   * enabled.
   *
   * @defaultValue `"top"`
   */
  pagePosition?: AppBarPosition;

  /**
   * Set this to `true` to remove the box-shadow.
   *
   * @defaultValue `false`
   */
  disableElevation?: boolean;

  /**
   * The theme to apply to the `AppBar`.
   *
   * @defaultValue `"primary"`
   * @see {@link AppBarTheme}
   */
  theme?: AppBarTheme;

  /**
   * Set this to `true` to update the `AppBar` to have `flex-direction: column`.
   * This is useful when nesting `AppBar`s.
   *
   * @remarks \@since 6.0.0
   * @defaultValue `false`
   */
  stacked?: boolean;

  /**
   * @defaultValue `stacked ? "auto" : "normal"`
   * @see {@link AppBarHeight}
   */
  height?: AppBarHeight;

  /**
   * Set this to `true` if the app bar's positioning and width should be
   * changed whenever the scrollbar is visible on the page. This defaults to
   * `true` when the {@link position} prop is `true` so that once dialogs and menus
   * become visible the contents in the app bar do not need to be repainted.
   *
   * @remarks \@since 6.0.0
   * @defaultValue `position === "fixed"`
   */
  scrollbarOffset?: boolean;
}

/**
 * Apply the `className`s for a tree component. This will be type-safe if using
 * typescript.
 *
 * @remarks \@since 6.0.0
 */
export function appBar(options: AppBarClassNameOptions = {}): string {
  const {
    className,
    height = "normal",
    theme = "primary",
    stacked = false,
    position = "static",
    pagePosition = "top",
    scrollbarOffset = position === "fixed",
    disableElevation = false,
  } = options;

  return cnb(
    styles({
      [theme]: theme !== "clear",
      [height]: height !== "normal",
      fixed: position !== "static",
      sticky: position === "sticky",
      stacked,
      [pagePosition]: position !== "static",
      elevated: position !== "static" && !disableElevation,
      "scrollbar-offset": scrollbarOffset,
      "static-scrollbar-offset": position === "static" && scrollbarOffset,
    }),
    className
  );
}

/** @remarks \@since 6.0.0 */
export type CustomAppBarComponent = ElementType<
  HTMLAttributes<HTMLDivElement> & {
    ref?: Ref<HTMLDivElement>;
    className: string;
  }
>;

/**
 * @remarks
 * \@since 6.0.0 Renamed the `component` prop to `as` and updated the default
 * value to be a `"div"` when `fixed` is `false`.
 * \@since 6.0.0 Renamed the `fixedElevation` prop to `disableFixedElevation` to
 * match naming conventions when a feature is enabled by default.
 * \@since 6.0.0 Removed the `inheritColor` and `flexWrap` props since they are
 * no longer required.
 * \@since 6.0.0 Removed the `fixed` prop in favor of the new `position` prop
 * which enables position `fixed` or `sticky` behavior.
 * \@since 6.0.0 Added the {@link stacked} and {@link scrollbarOffset} props.
 */
export interface AppBarProps
  extends HTMLAttributes<HTMLDivElement>,
    AppBarClassNameOptions {
  /** @defaultValue `fixed ? "header" : "div"` */
  as?: CustomAppBarComponent;
}

/**
 * **Server Component**
 *
 * @example
 * Simple Example
 * ```tsx
 * import { AppBar, AppBarTitle, Button } from "@react-md/core";
 * import MenuIcon from "@react-md/material-icons/MenuIcon";
 * import MoreVertIcon from "@react-md/material-icons/MoreVertIcon";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <AppBar>
 *       <Button aria-label="Navigation"><MenuIcon /></Button>
 *       <AppBarTitle>My Main Title</AppBarTitle>
 *       <Button aria-label="Options"><MoreVertIcon /></Button>
 *     </AppBar>
 *   );
 * }
 * ```
 *
 * @remarks
 * \@since 6.0.0 The `AppBar` was updated to use `gap` for spacing
 * instead of requiring the `AppBarNav` and `AppBarAction` components.
 */
export const AppBar = forwardRef<HTMLDivElement, AppBarProps>(
  function AppBar(props, ref) {
    const {
      className,
      theme,
      stacked,
      height = stacked ? "auto" : "normal",
      position,
      pagePosition,
      scrollbarOffset,
      disableElevation,
      as: Component = position ? "header" : "div",
      children,
      ...remaining
    } = props;

    return (
      <Component
        {...remaining}
        className={appBar({
          theme,
          position,
          pagePosition,
          disableElevation,
          stacked,
          height,
          className,
          scrollbarOffset,
        })}
        ref={ref}
      >
        {children}
      </Component>
    );
  }
);
