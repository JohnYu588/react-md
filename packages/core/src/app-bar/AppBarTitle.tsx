import { forwardRef } from "react";

import { type TextOverflow } from "../cssUtils.js";
import { Typography, type TypographyProps } from "../typography/Typography.js";
import { type TypographyType } from "../typography/typographyStyles.js";
import { type AppBarTitleClassNameOptions, appBarTitle } from "./styles.js";

/**
 * @since 6.0.0 The `keyline` prop was changed from a boolean to a type
 * union of different keylines: {@link AppBarTitleKeyline}
 * @since 6.0.0 The `noWrap` prop was removed in favor of the `textOverflow`
 * prop inherited through the base `Typography` component. The `textOverflow`
 * will default to `"ellipsis"` which is new as well.
 */
export interface AppBarTitleProps
  extends TypographyProps,
    AppBarTitleClassNameOptions {
  /** @defaultValue `"headline-6"` */
  type?: TypographyType;

  /**
   * @defaultValue `"ellipsis"`
   */
  textOverflow?: TextOverflow;
}

/**
 * @example Updating the Keyline
 * ```tsx
 * import { AppBar } from "@react-md/core/app-bar/AppBar";
 * import { AppBarTitle } from "@react-md/core/app-bar/AppBarTitle";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <AppBar>
 *       <AppBarTitle keyline="nav">
 *         Offset as if there was a nav button to the left
 *       </AppBarTitle>
 *     </AppBar>
 *   );
 * }
 * ```
 *
 * @since 6.0.0 The `keyline` prop was changed from a boolean to a type
 * union of different keylines: {@link AppBarTitleKeyline}
 * @since 6.0.0 The `noWrap` prop was removed in favor of the `textOverflow`
 * prop inherited through the base `Typography` component. The `textOverflow`
 * will default to `"ellipsis"` which is new as well.
 */
export const AppBarTitle = forwardRef<HTMLHeadingElement, AppBarTitleProps>(
  function AppBarTitle(props, ref) {
    const {
      type = "headline-6",
      keyline = "small",
      children,
      className,
      textOverflow = "ellipsis",
      ...remaining
    } = props;
    return (
      <Typography
        {...remaining}
        ref={ref}
        type={type}
        className={appBarTitle({
          keyline,
          className,
        })}
        textOverflow={textOverflow}
      >
        {children}
      </Typography>
    );
  }
);
