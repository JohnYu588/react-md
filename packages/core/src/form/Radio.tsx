"use client";
import { forwardRef } from "react";
import type { RadioProps } from "./InputToggle.js";
import { InputToggle } from "./InputToggle.js";

/**
 * **Client Component**
 *
 * You'll generally want to use the `useRadioGroup` hook along with this
 * component.
 *
 * @example
 * Simple Example
 * ```tsx
 * const { value, getRadioProps } = useRadioGroup({ name: "group" });
 *
 * return (
 *   <>
 *     <Radio {...getRadioProps("a")} label="First" />
 *     <Radio {...getRadioProps("b")} label="Second" />
 *     <Radio {...getRadioProps("c")} label="Third" />
 *   </>
 * );
 * ```
 */
export const Radio = forwardRef<HTMLInputElement, RadioProps>(
  function Radio(props, ref) {
    return <InputToggle {...props} ref={ref} type="radio" />;
  }
);
