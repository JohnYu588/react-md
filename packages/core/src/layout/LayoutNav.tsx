"use client";

import { cnb } from "cnbuilder";
import { type HTMLAttributes, type ReactNode, forwardRef } from "react";

import {
  type BaseSheetClassNameOptions,
  DEFAULT_SHEET_CLASSNAMES,
  DEFAULT_SHEET_TIMEOUT,
  sheet,
} from "../sheet/styles.js";
import {
  type CSSTransitionClassNames,
  type TransitionActions,
  type TransitionCallbacks,
  type TransitionTimeout,
} from "../transition/types.js";
import { useCSSTransition } from "../transition/useCSSTransition.js";
import { layoutNav } from "./layoutNavStyles.js";

/**
 * @since 6.0.0
 */
export interface LayoutNavProps
  extends HTMLAttributes<HTMLDivElement>,
    BaseSheetClassNameOptions,
    TransitionCallbacks,
    TransitionActions {
  children: ReactNode;

  /**
   * The component to render as.
   *
   * @defaultValue `"nav"`
   */
  as?: "nav" | "div";

  /**
   * Set this to `true` to display the navigation.
   */
  expanded: boolean;

  /**
   * Set this to `true` to force the navigation to appear below the fixed app
   * bar.
   *
   * @defaultValue `false`
   */
  appBarOffset?: boolean;

  /** @defaultValue {@link DEFAULT_SHEET_TIMEOUT} */
  timeout?: TransitionTimeout;
  /** @defaultValue {@link DEFAULT_SHEET_CLASSNAMES} */
  classNames?: CSSTransitionClassNames;
}

/**
 * **Client Component**
 *
 * This component is used to render a persistent navigation on the left of the
 * page. There is some built-in functionality to support resizing with the
 * `LayoutWindowSplitter` and expandable layouts. Do not use this component for
 * temporary navigation and instead use a `Sheet`.
 *
 * @example Full Height Layout
 * ```tsx
 * "use client";
 * import { LayoutAppBar } from "@react-md/core/layout/LayoutAppBar";
 * import { LayoutNav } from "@react-md/core/layout/LayoutNav";
 * import { Main } from "@react-md/core/layout/Main";
 * import { useHorizontalLayoutTransition } from "@react-md/core/layout/useHorizontalLayoutTransition";
 * import type { ReactElement, PropsWithChildren } from "react";
 *
 * function Layout({ children }: PropsWithChildren): ReactElement {
 *   const { elementProps } = useHorizontalLayoutTransition({
 *     transitionIn: true,
 *   });
 *
 *   return (
 *     <>
 *       <LayoutAppBar {...elementProps}>
 *         <YourAppBarContent />
 *       </LayoutAppBar>
 *       <LayoutNav expanded>
 *         <YourNavigationComponent />
 *       </LayoutNav>
 *       <Main navOffset appBarOffset {...elementProps}>
 *         {children}
 *       </Main>
 *     </>
 *   );
 * }
 * ```
 *
 * @example Toggleable Layout
 * ```tsx
 * "use client";
 * import { LayoutAppBar } from "@react-md/core/layout/LayoutAppBar";
 * import { Main } from "@react-md/core/layout/Main";
 * import { useHorizontalLayoutTransition } from "@react-md/core/layout/useHorizontalLayoutTransition";
 * import { useToggle } from "@react-md/core/useToggle";
 * import MenuIcon from "@react-md/material-icons/MenuIcon";
 * import { cnb } from "cnbuilder";
 * import type { ReactElement, PropsWithChildren } from "react";
 *
 * function Layout({ children }: PropsWithChildren): ReactElement {
 *   const { toggled: expanded, toggle } = useToggle();
 *   const { elementProps } = useHorizontalLayoutTransition({
 *     transitionIn: expanded,
 *   });
 *
 *   return (
 *     <>
 *       <LayoutAppBar {...elementProps}>
 *         <Button
 *           aria-label="Navigation"
 *           onClick={toggle}
 *           buttonType="icon"
 *         >
 *           <MenuIcon />
 *         </Button>
 *         <YourAppBarContent />
 *       </LayoutAppBar>
 *       <LayoutNav expanded>
 *         <YourNavigationComponent />
 *       </LayoutNav>
 *       <Main navOffset={expanded} appBarOffset {...elementProps}>
 *         {children}
 *       </Main>
 *     </>
 *   );
 * }
 * ```
 *
 * @since 6.0.0
 */
export const LayoutNav = forwardRef<HTMLDivElement, LayoutNavProps>(
  function LayoutNav(props, ref) {
    const {
      as: Component = "nav",
      "aria-labelledby": ariaLabelledBy,
      "aria-label": ariaLabel = Component == "nav" && !ariaLabelledBy
        ? "Navigation"
        : undefined,
      expanded,
      children,
      className,
      timeout = DEFAULT_SHEET_TIMEOUT,
      classNames = DEFAULT_SHEET_CLASSNAMES,
      appear,
      enter,
      exit,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExited,
      onExiting,
      appBarOffset,
      ...remaining
    } = props;
    const { elementProps } = useCSSTransition({
      nodeRef: ref,
      timeout,
      className: cnb(
        layoutNav({ appBarOffset }),
        sheet({
          className,
          horizontalSize: "none",
          disableOverlay: true,
        })
      ),
      classNames,
      enter,
      exit,
      appear,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExited,
      onExiting,
      exitedHidden: true,
      transitionIn: expanded,
    });

    return (
      <Component
        aria-label={ariaLabel}
        aria-labelledby={ariaLabelledBy}
        {...remaining}
        {...elementProps}
      >
        {children}
      </Component>
    );
  }
);
