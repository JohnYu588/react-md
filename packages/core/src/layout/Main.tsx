"use client";
import type { ElementType, HTMLAttributes, ReactNode, Ref } from "react";
import { forwardRef } from "react";
import { useEnsuredId } from "../useEnsuredId.js";
import type { MainClassNameOptions } from "./mainStyles.js";
import { main } from "./mainStyles.js";
import { useMainTabIndex } from "./useMainTabIndex.js";

/**
 * @remarks \@since 6.0.0
 */
export type CustomMainElement = ElementType<
  HTMLAttributes<HTMLElement> & {
    ref: Ref<HTMLElement>;
    className?: string;
    tabIndex?: number;
  }
>;

/**
 * @remarks \@since 6.0.0
 */
export interface MainProps
  extends HTMLAttributes<HTMLElement>,
    MainClassNameOptions {
  /**
   * @defaultValue `"main"`
   */
  as?: CustomMainElement;
  children: ReactNode;
}

/**
 * **Client Component**
 *
 * This component is really only used to dynamically set the `tabIndex` to `-1`
 * while using a keyboard for the `SkipToMainContent` component's focus behavior
 * to work correctly. If you don't need that functionality, use the {@link main}
 * style utility instead.
 *
 * @example
 * Styles Only
 * ```tsx
 * import { main as mainStyles } from "@react-md/core":
 *
 * function MyCustomMainElement({ children }) {
 *   return (
 *     <main
 *       className={mainStyles({
 *         navOffset: true,
 *         appBarOffset: true,
 *       })}
 *     >
 *       {children}
 *     </main>
 *   );
 * }
 * ```
 *
 * @remarks \@since 6.0.0 Renamed from `LayoutMain` removed a lot of
 * functionality to keep this component simple.
 */
export const Main = forwardRef<HTMLElement, MainProps>(
  function Main(props, ref) {
    const {
      as: Component = "main",
      id: propId,
      className,
      children,
      tabIndex: propTabIndex,
      navOffset,
      appBarOffset,
      ...remaining
    } = props;
    const id = useEnsuredId(propId, "main");
    const tabIndex = useMainTabIndex(propTabIndex);

    return (
      <Component
        {...remaining}
        id={id}
        ref={ref}
        className={main({ navOffset, appBarOffset, className })}
        tabIndex={tabIndex}
      >
        {children}
      </Component>
    );
  }
);
