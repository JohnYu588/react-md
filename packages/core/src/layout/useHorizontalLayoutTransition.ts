"use client";
import { cnb } from "cnbuilder";
import { DEFAULT_SHEET_TIMEOUT } from "../sheet/Sheet.js";
import {
  type CSSTransitionClassNames,
  type CSSTransitionHookReturnValue,
  type PreconfiguredCSSTransitionOptions,
  type TransitionTimeout,
} from "../transition/types.js";
import { useCSSTransition } from "../transition/useCSSTransition.js";

/**
 * @since 6.0.0
 */
export const DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES = {
  appearDone: "rmd-layout-h--active",
  enter: "rmd-layout-h--enter",
  enterActive: "rmd-layout-h--active",
  enterDone: "rmd-layout-h--active",
  exit: "rmd-layout-h--exit",
} satisfies Readonly<CSSTransitionClassNames>;

/**
 * @since 6.0.0
 */
export interface HorizontalLayoutTransitionOptions
  extends PreconfiguredCSSTransitionOptions<HTMLElement> {
  /**
   * @see {@link DEFAULT_SHEET_TIMEOUT}
   * @defaultValue `DEFAULT_SHEET_TIMEOUT`
   */
  timeout?: TransitionTimeout;
  /**
   * @see {@link DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES}
   * @defaultValue `DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES`
   */
  classNames?: CSSTransitionClassNames;
}

/**
 * @example
 * ```tsx
 * import { LayoutAppBar, Main, useHorizontalLayoutTransition } from "@react-md/core";
 * import type { ReactElement, ReactNode } from "react";
 *
 * interface Props {
 *   title: ReactNode;
 *   children: ReactNode;
 * }
 *
 * export default function Example({ title, children }: Props): ReactElement {
 *   const { toggled: staticNavExpanded, toggle: toggleStaticNav } = useToggle();
 *   const { elementProps, className } = useHorizontalLayoutTransition({
 *     transitionIn: staticNavExpanded,
 *   });
 *
 *   return (
 *     <>
 *       <LayoutAppBar className={className}>
 *         <Button
 *           aria-label="Navigation"
 *           buttonType="icon"
 *           onClick={toggleStaticNav}
 *         >
 *           <MenuOutlinedIcon />
 *         '</Button>
 *         {title}
 *       </LayoutAppBar>
 *       <Main {...elementProps}>
 *         {children}
 *       </Main>
 *     </>
 *   );
 * }
 * ```
 *
 * @since 6.0.0
 */
export function useHorizontalLayoutTransition<
  E extends HTMLElement = HTMLElement,
>(options: HorizontalLayoutTransitionOptions): CSSTransitionHookReturnValue<E> {
  return useCSSTransition({
    timeout: DEFAULT_SHEET_TIMEOUT,
    classNames: DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES,
    ...options,
    className: cnb("rmd-layout-h", options.className),
  });
}
