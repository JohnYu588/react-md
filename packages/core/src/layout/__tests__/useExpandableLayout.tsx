import { beforeEach, describe, expect, it } from "@jest/globals";
import { type ReactNode } from "react";

import { AppBar } from "../../app-bar/AppBar.js";
import { AppBarTitle } from "../../app-bar/AppBarTitle.js";
import { Button } from "../../button/Button.js";
import { Form } from "../../form/Form.js";
import { Radio } from "../../form/Radio.js";
import { useRadioGroup } from "../../form/useRadioGroup.js";
import { useAppSize } from "../../media-queries/AppSizeProvider.js";
import { Sheet } from "../../sheet/Sheet.js";
import {
  matchDesktop,
  matchPhone,
  matchTablet,
  rmdRender,
  screen,
  userEvent,
  waitFor,
  within,
} from "../../test-utils/index.js";
import { spyOnMatchMedia } from "../../test-utils/jest-globals/index.js";
import { TRANSITION_CONFIG } from "../../transition/config.js";
import { isElementVisible } from "../../utils/isElementVisible.js";
import { LayoutNav } from "../LayoutNav.js";
import { Main } from "../Main.js";
import {
  type ExpandableLayoutOptions,
  useExpandableLayout,
} from "../useExpandableLayout.js";
import { DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES } from "../useHorizontalLayoutTransition.js";

const OFFSCREEN_CLASS = "rmd-sheet--offscreen";
const EXIT_CLASS = "rmd-sheet--exit";
const {
  enter: ENTER_H_CLASS,
  enterDone: ENTER_H_DONE_CLASS,
  exit: EXIT_H_CLASS,
} = DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES;

interface LayoutProps extends Omit<ExpandableLayoutOptions, "pathname"> {
  children?: ReactNode;
}

function Layout(props: LayoutProps) {
  const { children = "Hello, world!", ...options } = props;

  const { getRadioProps, value: pathname } = useRadioGroup<string>({
    defaultValue: "/",
    name: "pathname",
  });
  const {
    temporary,
    appBarProps,
    mainProps,
    navToggleProps,
    temporaryNavProps,
    expandableNavProps,
  } = useExpandableLayout({ pathname, ...options });
  const { isTablet, isLandscape, isDesktop } = useAppSize();
  let layout: "phone" | "tablet" | "landscapeTablet" | "desktop" = "phone";
  if (isTablet) {
    layout = isLandscape ? "landscapeTablet" : "tablet";
  } else if (isDesktop) {
    layout = "desktop";
  }

  return (
    <>
      <AppBar {...appBarProps}>
        <Button {...navToggleProps} />
        <AppBarTitle>Hello, world!</AppBarTitle>
      </AppBar>
      <LayoutNav {...expandableNavProps}>
        {/* pretend like these are navigation items */}
        <Form>
          <Radio {...getRadioProps("/")} label="Home" />
          <Radio {...getRadioProps("/1")} label="Route 1" />
          <Radio {...getRadioProps("/2")} label="Route 2" />
        </Form>
      </LayoutNav>
      {temporary && (
        <Sheet {...temporaryNavProps}>
          {/* pretend like these are navigation items */}
          <Form>
            <Radio {...getRadioProps("/")} label="Home" />
            <Radio {...getRadioProps("/1")} label="Route 1" />
            <Radio {...getRadioProps("/2")} label="Route 2" />
          </Form>
        </Sheet>
      )}
      <Main {...mainProps}>
        {children}
        <div data-testid="layout">{layout}</div>
      </Main>
    </>
  );
}

describe("useExpandableLayout", () => {
  beforeEach(() => {
    TRANSITION_CONFIG.disabled = false;
  });

  it("should allow you to create an expandable layout quickly", async () => {
    const user = userEvent.setup();
    rmdRender(<Layout />);

    const layout = screen.getByTestId("layout");
    const appBar = screen.getByRole("banner");
    const main = screen.getByRole("main");
    const navToggle = screen.getByRole("button", { name: "Navigation" });
    const nav = screen.getByRole("navigation", { name: "Navigation" });

    expect(layout).toHaveTextContent("desktop");
    expect(isElementVisible(nav)).toBe(false);
    expect(nav).toHaveClass(OFFSCREEN_CLASS);
    expect(appBar).toMatchSnapshot();
    expect(main).toMatchSnapshot();
    expect(() => screen.getByRole("dialog")).toThrow();

    await user.click(navToggle);
    expect(isElementVisible(nav)).toBe(true);
    expect(main).toHaveClass(ENTER_H_CLASS);
    await waitFor(() => {
      expect(nav).not.toHaveClass(OFFSCREEN_CLASS);
    });
    expect(main).toHaveClass(ENTER_H_DONE_CLASS);

    await user.click(screen.getByRole("radio", { name: "Route 2" }));
    expect(nav).not.toHaveClass(EXIT_CLASS);
    expect(nav).toMatchSnapshot();

    await user.click(screen.getByRole("radio", { name: "Route 1" }));
    expect(nav).not.toHaveClass(EXIT_CLASS);
  });

  it("should not transition while changing between temporary and expanded layouts", async () => {
    const matchMediaSpy = spyOnMatchMedia();
    const user = userEvent.setup();
    rmdRender(<Layout />);

    const layout = screen.getByTestId("layout");
    const navToggle = screen.getByRole("button", { name: "Navigation" });
    const expandableNav = screen.getByRole("navigation", {
      name: "Navigation",
    });

    expect(layout).toHaveTextContent("desktop");
    expect(isElementVisible(expandableNav)).toBe(false);
    await user.click(navToggle);
    await waitFor(() => {
      expect(isElementVisible(expandableNav)).toBe(true);
    });
    await waitFor(() => {
      expect(expandableNav).not.toHaveClass(OFFSCREEN_CLASS);
    });

    matchMediaSpy.changeViewport(matchPhone);
    expect(layout).toHaveTextContent("phone");
    // note: this is really hidden due to CSS.
    expect(expandableNav).toBeInTheDocument();

    await user.click(navToggle);
    let temporaryNav = screen.getByRole("dialog", { name: "Navigation" });
    await waitFor(() => {
      expect(temporaryNav).not.toHaveClass(OFFSCREEN_CLASS);
    });
    expect(expandableNav).toBeInTheDocument();

    matchMediaSpy.changeViewport(matchDesktop);
    expect(expandableNav).toBeInTheDocument();
    expect(temporaryNav).not.toBeInTheDocument();

    matchMediaSpy.changeViewport(matchPhone);
    temporaryNav = screen.getByRole("dialog", { name: "Navigation" });
    expect(expandableNav).toBeInTheDocument();
    expect(temporaryNav).toBeInTheDocument();
    expect(temporaryNav).not.toHaveClass(OFFSCREEN_CLASS);

    await user.click(
      within(temporaryNav).getByRole("radio", { name: "Route 2" })
    );
    await waitFor(() => {
      expect(temporaryNav).not.toBeInTheDocument();
    });

    // the temporary navigation SHOULD close if the user resizes to desktop and changes routes
    await user.click(navToggle);
    temporaryNav = screen.getByRole("dialog", { name: "Navigation" });
    expect(temporaryNav).toHaveClass(OFFSCREEN_CLASS);
    await waitFor(() => {
      expect(temporaryNav).not.toHaveClass(OFFSCREEN_CLASS);
    });

    matchMediaSpy.changeViewport(matchDesktop);
    expect(temporaryNav).not.toBeInTheDocument();

    await user.click(screen.getByRole("radio", { name: "Home" }));
    expect(isElementVisible(expandableNav)).toBe(true);

    matchMediaSpy.changeViewport(matchPhone);
    expect(() => screen.getByRole("dialog")).toThrow();
  });

  it("should allow for the navigation to be visible by default", () => {
    rmdRender(<Layout defaultExpanded />);

    expect(screen.getByRole("navigation")).toBeInTheDocument();
  });

  it("should transition the app bar along with the main element for full height navigation layouts", async () => {
    const user = userEvent.setup();
    rmdRender(<Layout fullHeightNav />);

    const appBar = screen.getByRole("banner");
    const main = screen.getByRole("main");
    const navToggle = screen.getByRole("button", { name: "Navigation" });

    expect(appBar).not.toHaveClass(
      DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES.enter
    );
    expect(main).not.toHaveClass(
      DEFAULT_HORIZONTAL_LAYOUT_TRANSITION_CLASSNAMES.enter
    );

    await user.click(navToggle);
    expect(appBar).toHaveClass(ENTER_H_CLASS);
    expect(main).toHaveClass(ENTER_H_CLASS);
    await waitFor(() => {
      expect(appBar).not.toHaveClass(ENTER_H_CLASS);
    });
    await waitFor(() => {
      expect(main).not.toHaveClass(ENTER_H_CLASS);
    });
    expect(appBar).toHaveClass(ENTER_H_DONE_CLASS);
    expect(main).toHaveClass(ENTER_H_DONE_CLASS);

    await user.click(navToggle);
    expect(appBar).not.toHaveClass(ENTER_H_CLASS);
    expect(main).not.toHaveClass(ENTER_H_CLASS);
    expect(appBar).toHaveClass(EXIT_H_CLASS);
    expect(main).toHaveClass(EXIT_H_CLASS);
    await waitFor(() => {
      expect(appBar).not.toHaveClass(EXIT_H_CLASS);
    });
    await waitFor(() => {
      expect(main).not.toHaveClass(EXIT_H_CLASS);
    });
  });

  it("should allow the temporary layout to be used until desktop", () => {
    const matchMediaSpy = spyOnMatchMedia(matchPhone);
    rmdRender(<Layout temporaryUntil="desktop" defaultVisible />);

    const layout = screen.getByTestId("layout");
    const expandableNav = screen.getByRole("navigation", {
      name: "Navigation",
    });
    const temporaryNav = screen.getByRole("dialog", { name: "Navigation" });

    expect(layout).toHaveTextContent("phone");
    expect(isElementVisible(expandableNav)).toBe(false);
    expect(temporaryNav).toBeInTheDocument();

    matchMediaSpy.changeViewport(matchTablet);
    expect(temporaryNav).toBeInTheDocument();

    matchMediaSpy.changeViewport(matchDesktop);
    expect(temporaryNav).not.toBeInTheDocument();

    matchMediaSpy.mockRestore();
  });

  it("should allow for a static full height nav which will be hidden based on media queries", () => {
    rmdRender(<Layout fullHeightNav="static" defaultVisible />);
    const navToggle = screen.getByRole("button", { name: "Navigation" });
    expect(navToggle).toHaveClass("rmd-layout-nav-toggle");
  });
});
