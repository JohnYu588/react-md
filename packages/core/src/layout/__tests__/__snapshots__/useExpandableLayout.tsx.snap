// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useExpandableLayout should allow you to create an expandable layout quickly 1`] = `
<header
  class="rmd-app-bar rmd-app-bar--primary rmd-app-bar--fixed rmd-app-bar--top rmd-app-bar--elevated rmd-app-bar--scrollbar-offset"
>
  <button
    aria-label="Navigation"
    class="rmd-button rmd-button--icon"
    type="button"
  >
    <span
      aria-hidden="true"
      class="rmd-icon rmd-icon--font material-icons"
    >
      menu
    </span>
  </button>
  <h6
    class="rmd-typography rmd-typography--headline-6 rmd-app-bar-title rmd-app-bar-title--no-wrap"
  >
    Hello, world!
  </h6>
</header>
`;

exports[`useExpandableLayout should allow you to create an expandable layout quickly 2`] = `
<main
  class="rmd-layout-main rmd-layout-main--offset-v rmd-layout-h"
  id="main-:r3:"
>
  Hello, world!
  <div
    data-testid="layout"
  >
    desktop
  </div>
</main>
`;

exports[`useExpandableLayout should allow you to create an expandable layout quickly 3`] = `
<nav
  aria-label="Navigation"
  class="rmd-layout-nav rmd-layout-nav--offset-v rmd-sheet rmd-sheet--horizontal rmd-sheet--left"
>
  <form>
    <label
      class="rmd-label rmd-label--gap rmd-label--reversed"
    >
      Home
      <span
        class="rmd-radio rmd-input-toggle rmd-input-toggle--em rmd-input-toggle--normal"
      >
        <span
          aria-hidden="true"
          class="rmd-icon rmd-icon--font material-icons"
        >
          radio_button_unchecked
        </span>
        <input
          checked=""
          class="rmd-hidden-input"
          id="radio-:r0:"
          name="pathname"
          type="radio"
          value="/"
        />
      </span>
    </label>
    <label
      class="rmd-label rmd-label--gap rmd-label--reversed"
    >
      Route 1
      <span
        class="rmd-radio rmd-input-toggle rmd-input-toggle--em rmd-input-toggle--normal"
      >
        <span
          aria-hidden="true"
          class="rmd-icon rmd-icon--font material-icons"
        >
          radio_button_unchecked
        </span>
        <input
          class="rmd-hidden-input"
          id="radio-:r1:"
          name="pathname"
          type="radio"
          value="/1"
        />
      </span>
    </label>
    <label
      class="rmd-label rmd-label--gap rmd-label--reversed"
    >
      Route 2
      <span
        class="rmd-radio rmd-input-toggle rmd-input-toggle--em rmd-input-toggle--active rmd-input-toggle--normal"
      >
        <span
          aria-hidden="true"
          class="rmd-icon rmd-icon--font material-icons"
        >
          radio_button_checked
        </span>
        <input
          class="rmd-hidden-input"
          id="radio-:r2:"
          name="pathname"
          type="radio"
          value="/2"
        />
      </span>
    </label>
  </form>
</nav>
`;
